zilliqa/src/transaction.rs

```diff
@@ -41,6 +41,86 @@ use crate::{
     zq1_proto::{Code, Data, Nonce, ProtoTransactionCoreInfo},
 };
 
+/// Represents a validation result.
+/// This could be Result<String>, except that we would then return
+/// Result<Result<String>>, which would be confusing.
+/// The argument is a human-readable error message which can be returned to the
+/// user to indicate the problem with the transaction.
+#[derive(Debug, Copy, Clone)]
+pub enum ValidationOutcome {
+    Success,
+    /// Transaction input size exceeds configured limit - (size, limit)
+    TransactionInputSizeExceeded(usize, usize),
+    /// Transaction initcode size exceeds configured limit - (size, limit)
+    InitCodeSizeExceeded(usize, usize),
+    /// Gas limit exceeds block gas limit - (gas_limit, block_gas_limit)
+    BlockGasLimitExceeded(EvmGas, EvmGas),
+    /// Insufficient gas for zil transaction - (given, required)
+    InsufficientGasZil(ScillaGas, ScillaGas),
+    /// Insufficient gas for EVM transaction
+    InsufficientGasEvm(EvmGas, EvmGas),
+    /// Chain id was incorrect - (received, expected)
+    IncorrectChainId(u64, u64),
+    /// Insufficient funds in account - (txn_cost, account_balance)
+    InsufficientFunds(u128, u128),
+    /// Nonce too low - arg is the nonce we were expecting - (nonce, expected_nonce)
+    NonceTooLow(u64, u64),
+    /// Unrecognised type - not invocation, creation or transfer
+    UnknownTransactionType,
+}
+
+impl ValidationOutcome {
+    // I did try this with a vector, but sadly this involves too much
+    // trait magic to be convenient.
+    pub fn and_then<T>(&self, test: T) -> Result<ValidationOutcome>
+    where
+        T: FnOnce() -> Result<ValidationOutcome>,
+    {
+        if self.is_ok() {
+            test()
+        } else {
+            Ok(*self)
+        }
+    }
+
+    pub fn is_ok(&self) -> bool {
```

Comment:
I might also consider having your validation method return `Result<Result<(), ValidationError>, anyhow::Error>` (though note the last `anyhow::Error` is implied if you leave it out) which means callers can't accidentally forget to call `.is_ok()` on the result to decide what to do.

Reply:
I did consider that, but decided not to do if for fear of having callers become confused between the semantics of the outer and inner Result.